pkg/ring/kv/memberlist/metrics.go

```diff
@@ -0,0 +1,277 @@
+		"State information about memberlist cluster members",
+		[]string{"name", "address"}, nil)
+
+	m.watchPrefixDroppedNotifications = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: m.cfg.MetricsNamespace,
+		Subsystem: subsystem,
+		Name:      "watch_prefix_dropped_notifications",
+		Help:      "Number of dropped notifications in WatchPrefix function",
+	}, []string{"prefix"})
+
+	if m.cfg.MetricsRegisterer == nil {
+		return
+	}
+
+	// if registration fails, that's too bad, but don't panic
+	for _, c := range m.getAllMetrics() {
+		err := m.cfg.MetricsRegisterer.Register(c.(prometheus.Collector))
+		if err != nil {
+			level.Error(m.logger).Log("msg", "failed to register metrics", "err", err)
+		}
+	}
+
+	err := m.cfg.MetricsRegisterer.Register(m)
+	if err != nil {
+		level.Error(m.logger).Log("msg", "failed to register metrics", "err", err)
+	}
+}
+
+func (m *Client) unregisterMetrics() {
+	if m.cfg.MetricsRegisterer == nil {
+		return
+	}
+
+	for _, c := range m.getAllMetrics() {
+		m.cfg.MetricsRegisterer.Unregister(c.(prometheus.Collector))
+	}
+	m.cfg.MetricsRegisterer.Unregister(m)
+}
+
+func (m *Client) getAllMetrics() []prometheus.Collector {
+	return []prometheus.Collector{
+		m.numberOfReceivedMessages,
+		m.totalSizeOfReceivedMessages,
+		m.numberOfInvalidReceivedMessages,
+		m.numberOfBroadcastMessagesInQueue,
+		m.numberOfPushes,
+		m.numberOfPulls,
+		m.totalSizeOfPushes,
+		m.totalSizeOfPulls,
+		m.totalSizeOfBroadcastMessagesInQueue,
+		m.casAttempts,
+		m.casFailures,
+		m.casSuccesses,
+		m.watchPrefixDroppedNotifications,
+		m.memberlistMembersCount,
+		m.memberlistHealthScore,
+	}
+}
+
+// Describe returns prometheus descriptions via supplied channel
+func (m *Client) Describe(ch chan<- *prometheus.Desc) {
+	ch <- m.storeValuesDesc
+	ch <- m.storeSizesDesc
+	ch <- m.memberlistMembersInfoDesc
+}
+
+// Collect returns extra metrics via supplied channel
+func (m *Client) Collect(ch chan<- prometheus.Metric) {
+	ts := (*timestamps)(nil)
+	if obj, err := m.get(nodesTimestamps); err == nil {
+		if tsp, ok := obj.(*timestamps); ok {
+			ts = tsp
+		}
+	}
+
+	for _, n := range m.memberlist.Members() {
+		t := 0.0
+		if ts != nil {
+			ts, ok := ts.Timestamps[n.Name]
+			if ok {
+				t = float64(ts.UnixNano()) / float64(time.Second)
+			}
+		}
+
+		ch <- prometheus.MustNewConstMetric(m.memberlistMembersInfoDesc, prometheus.GaugeValue, t, n.Name, n.Address())
+	}
+
+	// Everything that needs KV Store lock is below here.
+	m.storeMu.Lock()
+	defer m.storeMu.Unlock()
+
+	ch <- prometheus.MustNewConstMetric(m.storeValuesDesc, prometheus.GaugeValue, float64(len(m.store)))
+
+	for k, v := range m.store {
+		ch <- prometheus.MustNewConstMetric(m.storeSizesDesc, prometheus.GaugeValue, float64(len(v.value)), k)
+	}
+}
+
+// Runs a goroutine that updates timestamp for this node every couple of seconds
+func (m *Client) updateNodeTimestamp() {
```

Comment:
I think we can assume that each node is gossiping every 15s (it's node update). And we will have a HTML page that shows how old each gossip was.

Reply:
"it's a node update" -- you mean ingesters, right? I wanted to cover more than just ingesters here... all components that read the ring.

When you're talking about HTML page, do you mean existing /ring page? Or something new?

By exporting the metrics, I wanted to have an ability to setup alerts if things go too wrong.

(I'm not saying this is good solution – it's not, just explaining the thinking behind it.)

Reply:
> By exporting the metrics, I wanted to have an ability to setup alerts if things go too wrong.

Could you list the alerts (what you want to alert for) here? Maybe we can come up with alternative ways to achieve that?

Reply:
I remember discussing this with you, but not if we have found a solution.

Alert we were thinking about is roughly this ... make sure that "now - last_heard_from < ~30s", where "last heard from" is every other node.

In practice, I think we need to make sure that all ingesters are up-to-date, and since we already have ingesters and their latest timestamp in the ring, we can use that. So what about creating a two metrics:

1) number of non-left ingesters in the ring
2) oldest timestamp of non-left ingester in the ring

If 1) is greater than 0, and 2) is greater than (say) 30 seconds, we can raise an alert. These two metrics would be exposed by all components that use the ring.

How does that sound?

Now this solution is fine (hopefully) for ring with ingesters, but if we reuse memberlist KV store for something else, we need new metrics for that.

Reply:
I've removed timestamps gossiping, and per-value codec. Need to add those two new metrics now.